src/src/Model/AbandonedCartCollection.php

```diff
@@ -0,0 +1,105 @@
+<?php
+/**
+ * 2024 Smaily
+ *
+ * NOTICE OF LICENSE
+ *
+ * Smaily for PrestaShop is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * any later version.
+ *
+ * Smaily for PrestaShop is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Smaily for PrestaShop. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @author    Smaily <info@smaily.com>
+ * @copyright 2024 Smaily
+ * @license   GPL3
+ */
+declare(strict_types=1);
+
+namespace PrestaShop\Module\SmailyForPrestaShop\Model;
+
+if (!defined('_PS_VERSION_')) {
+    exit;
+}
+
+class AbandonedCartCollection
+{
+    /**
+     * @var \Db;
+     */
+    private $db;
+
+    /**
+     * @var AbandonedCart[]
+     */
+    public $carts;
+
+    public function __construct()
+    {
+        $this->db = \Db::getInstance();
+        $this->carts = $this->getCarts();
+    }
+
+    /**
+     * Gets abandoned cart data from DB.
+     *
+     * @return AbandonedCart[] Abandoned carts array
+     */
+    private function getCarts()
+    {
+        $sql = new \DbQuery();
+        $sql->select('c.`id_cart`, c.`id_customer`, c.`date_upd`, cu.`firstname`, cu.`lastname`, cu.`email`');
+        $sql->from('cart', 'c');
+        $sql->leftJoin('orders', 'o', 'o.`id_cart` = c.`id_cart`');
+        $sql->rightJoin('customer', 'cu', 'cu.`id_customer` = c.`id_customer`');
+        $sql->leftOuterJoin('smaily_cart', 'sc', 'c.`id_cart` = sc.`id_cart`');
+        $sql->where('sc.`id_cart` IS NULL');
+        $sql->where('DATE_SUB(CURDATE(),INTERVAL 10 DAY) <= c.date_add');
+        $sql->where('o.`id_order` IS NULL');
+        $sql->groupBy('cu.`id_customer`');
+        $sql .= \Shop::addSqlRestriction(\Shop::SHARE_CUSTOMER, 'c');
```

Comment:
I think the internals somehow took care of this, but the query doesn't make sense that way. This works but not so sure why. I improved it and now use the same legacy method but integrate it into the `DbQuery` builder. https://github.com/sendsmaily/smaily-prestashop-module/pull/57/commits/c906154150898763fe6c430f145362e3e581795c

Reply:
Using `ltrim` to strip ` AND ` from the WHERE clause can have some ill effects, because `ltrim`'s second argument is a list of characters to remove, i.e. ` `, `A`, `N`, `D` will treated separately and all removed, meaning ` AAANNNDDD foo` would still be `foo`.

Reply:
I have been using this function wrong the whole time probably. https://github.com/sendsmaily/smaily-prestashop-module/pull/57/commits/234cf03ab95cef03b109ca14f66876c277449f8d

Reply:
The solution you have opted to works as long as there is only one WHERE clause returned by `\Shop::addSqlRestriction(...)`.

Meaning ` AND [shop-restriction]` works as expected, but ` AND restriction_1 = 'foo' AND restriction_2 = 'bar'` doesn't. The latter would yield in `restriction_1 = 'foo'  restriction_2 = 'bar'`, which isn't valid SQL WHERE clause.

Given the volatility of PrestaShop at the moment I would opt for a more foolproof method, like regular expression.

Reply:
There is a single `AND` statement inside the `addSqlRestriction` function. The logic hasn't been changed for 9 years and probably won't be touched again as the migration path doesn't allow this.

https://github.com/PrestaShop/PrestaShop/blob/6fe208192e08b33b02e1cda3aa46f41851c7cf6a/classes/shop/Shop.php#L1091-L1109

Further improvements (PrestaShop versions) should use `src/Adapter` or `src/Core` classes when they become ported. I think the regex here is overkill as we know the function internals.

Reply:
Assumption is the mother of all f*ckups. We can't be sure of the migration path PrestaShop needs to take and cannot be certain the internal implementation of the method won't change.

Current implementation means that abandoned cart stops working and we would need to go and fix it, if and when somebody changes the internal implementation. A simple regular expression of:

```php
preg_replace("/^\s+AND\s+/", "", \Shop::addSqlRestriction(...));
```

would only do what is needed and wouldn't break if the PrestaShop's internal implementation should change, i.e. when additional conditions are introduced or entirely changed.

Reply:
Okidok, let's use the regular expression https://github.com/sendsmaily/smaily-prestashop-module/pull/57/commits/0eebd84f5239f0ef437ffa3dee2c7a32a2e2264a

Reply:
Thank you!

A slight formatting issue was introduced though: double space between `=` and `preg_replace`.

https://github.com/sendsmaily/smaily-prestashop-module/commit/0eebd84f5239f0ef437ffa3dee2c7a32a2e2264a#diff-24da1b0fab83948b3ac3f39f1976ad9f181dbd75a5ca25a19f6889b8b1e91d40R58

Reply:
Fixed  https://github.com/sendsmaily/smaily-prestashop-module/pull/57/commits/60396e4ad713f9887623970c0c5489305ffa4034